opm/autodiff/ExtendedBlackoilModel_impl.hpp

```diff
@@ -0,0 +1,703 @@
+#include <opm/core/utility/ErrorMacros.hpp>
+#include <opm/core/utility/Exceptions.hpp>
+#include <opm/core/utility/Units.hpp>
+#include <opm/core/well_controls.h>
+#include <opm/core/utility/parameters/ParameterGroup.hpp>
+
+#include <cassert>
+#include <cmath>
+#include <iostream>
+#include <iomanip>
+#include <limits>
+
+namespace Opm {
+
+
+
+    namespace detail {
+
+        template <class PU>
+        int solventPos(const PU& pu)
+        {
+            const int maxnp = Opm::BlackoilPhases::MaxNumPhases;
+            int pos = 0;
+            for (int phase = 0; phase < maxnp; ++phase) {
+                if (pu.phase_used[phase]) {
+                    pos++;
+                }
+            }
+
+            return pos;
+        }
+
+    } // namespace detail
+
+
+
+    template <class Grid>
+    ExtendedBlackoilModel<Grid>::ExtendedBlackoilModel(const typename Base::ModelParameters&   param,
+                                                     const Grid&                             grid,
+                                                     const BlackoilPropsAdInterface&         fluid,
+                                                     const DerivedGeology&                   geo,
+                                                     const RockCompressibility*              rock_comp_props,
+                                                     const SolventPropsAdFromDeck&           solvent_props,
+                                                     const Wells*                            wells,
+                                                     const NewtonIterationBlackoilInterface& linsolver,
+                                                     const EclipseStateConstPtr              eclState,
+                                                     const bool                              has_disgas,
+                                                     const bool                              has_vapoil,
+                                                     const bool                              terminal_output,
+                                                     const bool                              has_solvent)
+        : Base(param, grid, fluid, geo, rock_comp_props, wells, linsolver,
+               eclState, has_disgas, has_vapoil, terminal_output),
+          has_solvent_(has_solvent),
+          solvent_pos_(detail::solventPos(fluid.phaseUsage())),
+          solvent_props_(solvent_props)
+    {
+        if (has_solvent_) {
+
+            // If deck has solvent, residual_ should contain solvent equation.
+            rq_.resize(fluid_.numPhases() + 1);
+            residual_.material_balance_eq.resize(fluid_.numPhases() + 1, ADB::null());
+            assert(solvent_pos_ == fluid_.numPhases());
+            if (has_vapoil_) {
+                OPM_THROW(std::runtime_error, "Solvent option only works with dead gas\n");
+            }
+        }
+    }
+
+
+
+    template <class Grid>
```

Comment:
Since this method (and also afterStep()) are unneeded in this class they should be removed.